apps/web/src/dialogs/command-palette/command-palette-dialog.tsx

```diff
@@ -0,0 +1,461 @@
+
+Copyright (C) 2023 Streetwriters (Private) Limited
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { debounce, toTitleCase } from "@notesnook/common";
+import { Box, Button, Flex, Text } from "@theme-ui/components";
+import React, { useEffect, useRef, useState } from "react";
+import Skeleton from "react-loading-skeleton";
+import { GroupedVirtuoso, GroupedVirtuosoHandle } from "react-virtuoso";
+import { db } from "../../common/db";
+import { BaseDialogProps, DialogManager } from "../../common/dialog-manager";
+import Dialog from "../../components/dialog";
+import Field from "../../components/field";
+import {
+  Cross,
+  Icon,
+  Notebook as NotebookIcon,
+  Note as NoteIcon,
+  Reminder as ReminderIcon,
+  Tag as TagIcon
+} from "../../components/icons";
+import { CustomScrollbarsVirtualList } from "../../components/list-container";
+import { hashNavigate, navigate } from "../../navigation";
+import { useEditorStore } from "../../stores/editor-store";
+import Config from "../../utils/config";
+import { commands as COMMANDS } from "./commands";
+
+interface Command {
+  id: string;
+  title: string;
+  highlightedTitle?: string;
+  type: "command" | "note" | "notebook" | "tag" | "reminder";
+  group: string;
+}
+
+type GroupedCommands = { group: Command["group"]; count: number }[];
+
+export const CommandPaletteDialog = DialogManager.register(
+  function CommandPaletteDialog(props: BaseDialogProps<boolean>) {
+    const [commands, setCommands] = useState<Command[]>(getDefaultCommands());
+    const [selected, setSelected] = useState(0);
+    const [query, setQuery] = useState(">");
+    const [loading, setLoading] = useState(false);
+    const virtuosoRef = useRef<GroupedVirtuosoHandle>(null);
+
+    useEffect(() => {
+      virtuosoRef.current?.scrollToIndex({
+        index: selected,
+        align: "end",
+        behavior: "auto"
+      });
+    }, [selected]);
+
+    async function onChange(e: React.ChangeEvent<HTMLInputElement>) {
+      setSelected(0);
+      const query = e.target.value;
+      setQuery(query);
+      !isCommandMode(query) && setLoading(true);
+      const res = await search(query);
+      const highlighted = db.lookup.fuzzy(
+        prepareQuery(query),
+        res.map((r) => ({
+          ...r,
+          highlightedTitle: r.title
+        })) ?? [],
+        "highlightedTitle",
```

Comment:
Why are we using a separate key for `highlightedTitle` instead of just replacing the `title`?

Reply:
Highlighting the title key would also add the highlighted title to recent commands in local storage

Reply:
Would be a good idea to add a comment for that.